src/util/mod.rs

```diff
@@ -34,6 +34,39 @@ impl ByteStr {
     }
 }
 
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub(crate) struct PercentDecodedByteStr(ByteStr);
+
+impl PercentDecodedByteStr {
+    pub(crate) fn new<S>(s: S) -> Option<Self>
+    where
+        S: AsRef<str>,
+    {
+        let byte_str = if let Ok(decoded) =
+            percent_encoding::percent_decode(s.as_ref().as_bytes()).decode_utf8()
+        {
+            ByteStr::new(decoded)
+        } else {
+            return None;
+        };
+
+        Some(Self(byte_str))
```

Comment:
This seems unnecessarily verbose with the "early" `return`, but I guess that's a just matter of taste.

Reply:
Sure